numba/np/npyfuncs.py

```diff
@@ -584,29 +585,42 @@ def np_complex_sign_impl(context, builder, sig, args):
     # equivalent to complex sign in NumPy's sign
     # but implemented via selects, balancing the 4 cases.
     _check_arity_and_homogeneity(sig, args, 1)
-    op = args[0]
-    ty = sig.args[0]
-    float_ty = ty.underlying_float
 
-    ZERO = context.get_constant(float_ty, 0.0)
-    ONE  = context.get_constant(float_ty, 1.0)
-    MINUS_ONE = context.get_constant(float_ty, -1.0)
-    NAN = context.get_constant(float_ty, float('nan'))
-    result = context.make_complex(builder, ty)
-    result.real = ZERO
-    result.imag = ZERO
-
-    cmp_sig = typing.signature(types.boolean, *[ty] * 2)
-    cmp_args = [op, result._getvalue()]
-    arg1_ge_arg2 = np_complex_ge_impl(context, builder, cmp_sig, cmp_args)
-    arg1_eq_arg2 = np_complex_eq_impl(context, builder, cmp_sig, cmp_args)
-    arg1_lt_arg2 = np_complex_lt_impl(context, builder, cmp_sig, cmp_args)
-
-    real_when_ge = builder.select(arg1_eq_arg2, ZERO, ONE)
-    real_when_nge = builder.select(arg1_lt_arg2, MINUS_ONE, NAN)
-    result.real = builder.select(arg1_ge_arg2, real_when_ge, real_when_nge)
-
-    return result._getvalue()
+    if numpy_version >= (2, 0):
+        # NumPy >= 2.0.0
+        def complex_sign(z):
+            abs = math.hypot(z.real, z.imag)
+            if abs == 0:
+                return 0 + 0j
+            else:
+                return z / abs
```

Comment:
Note: This sort-of follows the description from the docs of how complex sign should behave and will likely work reasonably for normal numbers, but will not replicate NumPy for special numbers like `inf` and `nan`. However, as noted OOB, it seems like in the case of special numbers, NumPy itself doesn't follow the array-api specification as is suggested it should do in the docs. I think we should leave this as-is until the issue is resolved with NumPy upstream.